javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -748,12 +771,12 @@ public boolean isEmpty() {
 
         @Override
         public boolean isLeaf() {
-            return children.isEmpty();
+            return size == 1;
         }
 
         @Override
         public int size() {
-            return 1 + children.foldLeft(0, (acc, child) -> acc + child.length());
+            return size;
```

Comment:
I'm not sure I can follow, it seems to me that the two calculations are exactly the same (recursive iteration in both cases).

* The current impl is faster, if you call `size` 0 times
* The new imp is faster if you call it `> 1` times

With lazy init would be the best of both worlds.

In the constructor (or rather after the field)
```Java
this.size = Lazy.of(() -> nodeCount());
```

and use it as:
```Java
public int size() { return size.get(); }
```